File: CallbackTechniquesTests/NotificationCenter/NotificationCenter.m
//
//  NotificationCenter.m
//  CallbackTechniques
//
//  Created by Alexey Naumov on 29/01/2017.
//  Copyright © 2017 Alexey Naumov. All rights reserved.
//

#import "NotificationCenter.h"

NSString *const TestNotification = @"test_notification";

@implementation ObjCNotificationCenterCaller

- (void)postNotification
{
  [[NSNotificationCenter defaultCenter] postNotificationName:TestNotification object:self];
}

@end

@implementation ObjCNotificationCenterCallee

- (void)subscribe
{
  [[NSNotificationCenter defaultCenter] addObserver:self selector:@selector(handleNotification:) name:TestNotification object:nil];
}

- (void)unsubscribe
{
  [[NSNotificationCenter defaultCenter] removeObserver:self name:TestNotification object:nil];
}

- (void)handleNotification:(NSNotification *)notification
{
  self.wasCalled = YES;
}

@end

File: CallbackTechniques/NotificationCenter/NotificationCenter.swift
//
//  NotificationCenter.swift
//  CallbackTechniques
//
//  Created by Alexey Naumov on 29/01/2017.
//  Copyright © 2017 Alexey Naumov. All rights reserved.
//

extension NSNotification.Name {
  static let Test = NSNotification.Name(rawValue: "test_notification")
}

class SwiftNotificationCenterCaller {
  func postNotification() {
    NotificationCenter.default.post(name: NSNotification.Name.Test, object: self)
  }
}

class SwiftNotificationCenterCallee {
  
  var wasCalled = false
  
  func subscribe() {
    NotificationCenter.default.addObserver(self, selector: #selector(handleNotification(notification:)), name: NSNotification.Name.Test, object: nil)
  }
  
  func unsubscribe() {
    NotificationCenter.default.removeObserver(self, name: NSNotification.Name.Test, object: nil)
  }
  
  @objc func handleNotification(notification: Notification) {
    wasCalled = true
  }
}
